Source/JavaScriptCore/inspector/agents/InspectorDebuggerAgent.cpp

```diff
@@ -661,6 +661,90 @@ Protocol::ErrorStringOr<void> InspectorDebuggerAgent::removeBreakpoint(const Pro
     return { };
 }
 
+static String functionName(JSC::CodeBlock* codeBlock)
+{
+    if (!codeBlock)
+        return nullString();
+
+    if (codeBlock->codeType() != JSC::CodeType::FunctionCode)
+        return nullString();
+
+    return JSC::jsCast<JSC::FunctionExecutable*>(codeBlock->ownerExecutable())->ecmaName().string();
+}
+
+static String functionName(JSC::CallFrame* callFrame)
+{
+    if (!callFrame)
+        return nullString();
+
+    if (auto* codeBlock = callFrame->codeBlock())
+        return functionName(codeBlock);
+
+    return nullString();
+}
+
+Protocol::ErrorStringOr<void> InspectorDebuggerAgent::addSymbolicBreakpoint(const String& symbol, std::optional<bool>&& caseSensitive, std::optional<bool>&& isRegex, RefPtr<JSON::Object>&& options)
+{
+    Protocol::ErrorString errorString;
+
+    auto breakpoint = debuggerBreakpointFromPayload(errorString, WTFMove(options));
+    if (!breakpoint)
+        return makeUnexpected(errorString);
+
+    {
+        SymbolicBreakpoint symbolicBreakpoint;
+        symbolicBreakpoint.symbol = symbol;
+        if (caseSensitive)
+            symbolicBreakpoint.caseSensitive = *caseSensitive;
+        if (isRegex)
+            symbolicBreakpoint.isRegex = *isRegex;
+        symbolicBreakpoint.specialBreakpoint = WTFMove(breakpoint);
+
+        if (!m_symbolicBreakpoints.appendIfNotContains(WTFMove(symbolicBreakpoint)))
+            return makeUnexpected("Symbolic breakpoint for given symbol, given caseSensitive, and given isRegex already exists"_s);
+    }
+
+    auto& symbolicBreakpoint = m_symbolicBreakpoints.last();
+
+    {
+        JSC::JSLockHolder locker(m_debugger.vm());
```

Comment:
It appears we must hold this lock to iterate code blocks, so can we just take the lock inside `forEachRegisteredCodeBlock` instead? Unless you think we will at some point start using the function in code that already holds the lock...

Aside: It's unfortunate there isn't currently a way (that I can find) to compile-time or runtime enforce a `JSLock`.

Reply:
according to @saambarati, we should also be holding the lock when calling `forEachRegisteredCodeBlock` too